File: fix_imbalance.py
#!/usr/bin/env python3
"""
fix_imbalance.py — Fix the AI / Intermediate imbalance for the next 14 days.

Rules applied:
  - Max 1 Advanced Intermediate per day
  - Target 2 Intermediate sessions per day (add if missing)
  - Never cancel an occurrence that has registered members
  - Vary AI start times across days (for pattern discovery)

Usage:
    venv/bin/python fix_imbalance.py             # dry run (shows what would change)
    venv/bin/python fix_imbalance.py --execute   # actually make the changes

    make fix-imbalance          # dry run
    make fix-imbalance-execute  # execute
"""

import argparse
import json
import logging
import sys
from datetime import date, datetime, timedelta
from pathlib import Path

from dotenv import load_dotenv
load_dotenv(Path(__file__).parent / ".env", override=True)

from cr_client import browser_session, fetch_schedule
from policy_loader import load_policy
from book_event import book_event, cancel_occurrence

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s  %(levelname)-7s  %(message)s",
    handlers=[logging.StreamHandler(sys.stdout)],
)
log = logging.getLogger(__name__)

POLICY        = load_policy()
APPROVED      = POLICY["approved_events"]          # str(id) -> {name, level}
COURTS        = POLICY["courts"]                   # str(id) -> {number, label}

# Approved event IDs by level
_BY_LEVEL = {info["level"]: eid for eid, info in APPROVED.items()}
AI_EVENT_ID   = int(_BY_LEVEL["Advanced Intermediate"])
INT_EVENT_ID  = int(_BY_LEVEL["Intermediate"])

# Court IDs in preference order (Court 4 preferred, then 1, 2, 3)
_COURT_IDS    = sorted(COURTS.keys(), key=lambda c: (COURTS[c]["number"] != 4, COURTS[c]["number"]))
PRIMARY_COURT_ID = int(_COURT_IDS[0])

# AI start-time rotation — vary across days to build pattern data
# Format: hour (24h). We rotate through these so each day gets a different slot.
AI_TIME_ROTATION = [9, 11, 13, 15, 17]   # 9am, 11am, 1pm, 3pm, 5pm

# Intermediate target and preferred times
INT_TARGET  = 2   # sessions per day
INT_TIMES   = [9, 11, 13, 15]  # candidate start hours


def _get_level(item: dict) -> str | None:
    eid = str(item.get("EventId", ""))
    return APPROVED[eid]["level"] if eid in APPROVED else None


def _court_free_at(items: list[dict], court_id: int, start: datetime, end: datetime) -> bool:
    """Return True if this court has no overlapping event."""
    for item in items:
        if str(item.get("CourtId") or "") != str(court_id):
            # Also check Courts field if CourtId not present
            courts_str = item.get("Courts", "")
            if str(court_id) not in courts_str:
                continue
        try:
            s = datetime.fromisoformat(item["StartDateTime"])
            e = datetime.fromisoformat(item["EndDateTime"])
        except Exception:
            continue
        if s < end and e > start:
            return False
    return True


def _find_free_court(items: list[dict], start: datetime, end: datetime) -> int | None:
    """Return first available court ID at this time, respecting the 3-court max."""
    occupied = set()
    for item in items:
        try:
            s = datetime.fromisoformat(item["StartDateTime"])
            e = datetime.fromisoformat(item["EndDateTime"])
        except Exception:
            continue
        if s < end and e > start:
            courts_str = str(item.get("Courts", ""))
            for cid in COURTS:
                if str(COURTS[cid]["number"]) in courts_str or cid in courts_str:
                    occupied.add(int(cid))

    # max 3 courts occupied at any time (hard constraint)
    if len(occupied) >= 3:
        return None

    for cid in _COURT_IDS:
        if int(cid) not in occupied:
            return int(cid)
    return None


def analyse_day(date_str: str, items: list[dict]) -> dict:
    """Return analysis of AI and Intermediate events for one day."""
    dt = datetime.strptime(date_str, "%m/%d/%Y")
    ai_events  = []
    int_events = []

    for item in items:
        level = _get_level(item)
        start = datetime.fromisoformat(item["StartDateTime"])
        if start.date() != dt.date():
            continue
        members = int(item.get("MembersCount") or 0)
        record = {
            "event_name":    item.get("EventName", ""),
            "start":         start,
            "end":           datetime.fromisoformat(item["EndDateTime"]),
            "courts":        item.get("Courts", ""),
            "members":       members,
            "occurrence_id": item.get("Id"),
            "event_id":      item.get("EventId"),
        }
        if level == "Advanced Intermediate":
            ai_events.append(record)
        elif level == "Intermediate":
            int_events.append(record)

    return {
        "date_str":   date_str,
        "dow":        dt.strftime("%A"),
        "ai_events":  sorted(ai_events,  key=lambda x: x["start"]),
        "int_events": sorted(int_events, key=lambda x: x["start"]),
        "all_items":  items,
    }


def plan_changes(analyses: list[dict]) -> list[dict]:
    """
    Given per-day analyses, produce a list of changes to make.
    Each change: {action: 'cancel'|'book', date_str, ...}
    """
    changes = []

    for i, day in enumerate(analyses):
        date_str  = day["date_str"]
        dow       = day["dow"]
        ai_events = day["ai_events"]
        int_events = day["int_events"]

        # ── AI: keep at most 1 ───────────────────────────────────────────────
        if len(ai_events) > 1:
            # Keep the one with members (or first if none), cancel rest
            with_members = [e for e in ai_events if e["members"] > 0]
            keep = with_members[0] if with_members else ai_events[0]
            for ev in ai_events:
                if ev is keep:
                    continue
                if ev["members"] > 0:
                    log.warning(
                        "  %s  AI at %s has %d members — SKIPPING cancel",
                        date_str, ev["start"].strftime("%-I:%M%p"), ev["members"],
                    )
                    continue
                changes.append({
                    "action":       "cancel",
                    "date_str":     date_str,
                    "dow":          dow,
                    "event_id":     int(ev["event_id"]),
                    "occurrence_id": int(ev["occurrence_id"]),
                    "description":  f"Cancel extra AI at {ev['start'].strftime('%-I:%M%p')} ({ev['courts']})",
                })

        # ── Intermediate: add up to target ───────────────────────────────────
        needed = INT_TARGET - len(int_events)
        if needed > 0:
            items_for_day = day["all_items"]
            # Try candidate start hours
            for hour in INT_TIMES:
                if needed <= 0:
                    break
                start_dt = datetime.strptime(date_str, "%m/%d/%Y").replace(hour=hour)
                end_dt   = start_dt.replace(hour=hour + 2)

                # Check operating window
                win = POLICY["operating_windows"]
                window = win["weekday"] if dow not in ("Saturday", "Sunday") else win["weekend"]
                win_start_h = int(window["start"].split(":")[0])
                win_end_h   = int(window["end"].split(":")[0])
                if hour < win_start_h or hour + 2 > win_end_h:
                    continue

                # Skip if this time already has an Intermediate session
                already = any(
                    abs((e["start"].hour * 60 + e["start"].minute) - hour * 60) < 120
                    for e in int_events
                )
                if already:
                    continue

                # Find a free court (respecting 3-court max)
                court_id = _find_free_court(items_for_day, start_dt, end_dt)
                if court_id is None:
                    continue

                changes.append({
                    "action":      "book",
                    "date_str":    date_str,
                    "dow":         dow,
                    "event_id":    INT_EVENT_ID,
                    "start_time":  start_dt.strftime("%-I:%M %p"),
                    "end_time":    end_dt.strftime("%-I:%M %p"),
                    "court_id":    court_id,
                    "court_num":   COURTS[str(court_id)]["number"],
                    "description": (
                        f"Book Intermediate {start_dt.strftime('%-I:%M%p')}–"
                        f"{end_dt.strftime('%-I:%M%p')} Court #{COURTS[str(court_id)]['number']}"
                    ),
                })
                # Add to items_for_day so subsequent slots respect this booking
                items_for_day.append({
                    "EventId":       str(INT_EVENT_ID),
                    "StartDateTime": start_dt.isoformat(),
                    "EndDateTime":   end_dt.isoformat(),
                    "Courts":        f"#{COURTS[str(court_id)]['number']}",
                    "CourtId":       str(court_id),
                    "MembersCount":  0,
                })
                needed -= 1

    return changes


def print_summary(analyses: list[dict], changes: list[dict]):
    """Print a human-readable summary."""
    print("\n" + "═" * 70)
    print("  CURRENT SCHEDULE — Advanced Intermediate & Intermediate")
    print("═" * 70)
    for day in analyses:
        ai  = day["ai_events"]
        i   = day["int_events"]
        if not ai and not i:
            continue
        def _ai_label(e):
            tag = "🔒" if e["members"] > 0 else str(e["members"])
            return f"{e['start'].strftime('%-I%p')}({tag})"
        ai_str = ", ".join(_ai_label(e) for e in ai) or "—"
        i_str = ", ".join(
            f"{e['start'].strftime('%-I%p')}({e['members']})"
            for e in i
        ) or "—"
        flag = "  ⚠" if len(ai) > 1 or len(i) < INT_TARGET else ""
        print(f"  {day['dow'][:3]} {day['date_str']}  AI:[{ai_str}]  I:[{i_str}]{flag}")

    print("\n" + "═" * 70)
    print(f"  PLANNED CHANGES  ({len(changes)} total)")
    print("═" * 70)
    if not changes:
        print("  No changes needed.")
    for c in changes:
        icon = "❌ CANCEL" if c["action"] == "cancel" else "✅ BOOK"
        print(f"  {icon}  {c['dow'][:3]} {c['date_str']}  {c['description']}")
    print()


def execute_changes(changes: list[dict], dry_run: bool):
    """Execute the planned changes against Court Reserve."""
    if not changes:
        log.info("Nothing to do.")
        return

    cancels = [c for c in changes if c["action"] == "cancel"]
    bookings = [c for c in changes if c["action"] == "book"]

    with browser_session(headless=False) as page:
        # ── Cancellations first ──────────────────────────────────────────────
        for c in cancels:
            log.info("Cancelling %s occ=%s on %s", c["description"], c["occurrence_id"], c["date_str"])
            result = cancel_occurrence(
                page          = page,
                event_id      = c["event_id"],
                occurrence_id = c["occurrence_id"],
                date          = c["date_str"],
                dry_run       = dry_run,
            )
            status = "✅ OK" if result["success"] else f"❌ FAILED: {result.get('error')}"
            print(f"  {status}  {c['description']}")

        # ── New bookings ─────────────────────────────────────────────────────
        for c in bookings:
            log.info("Booking %s on %s", c["description"], c["date_str"])
            result = book_event(
                page       = page,
                event_id   = c["event_id"],
                date       = c["date_str"],
                start_time = c["start_time"],
                end_time   = c["end_time"],
                court_id   = c["court_id"],
                dry_run    = dry_run,
            )
            status = "✅ OK" if result["success"] else f"❌ FAILED: {result.get('error')}"
            print(f"  {status}  {c['description']}")


def main():
    parser = argparse.ArgumentParser(description="Fix AI/Intermediate imbalance for the next 14 days")
    parser.add_argument("--execute", action="store_true",
                        help="Actually make the changes (default: dry run, show plan only)")
    parser.add_argument("--days", type=int, default=14,
                        help="Number of days ahead to check (default: 14)")
    args = parser.parse_args()

    dry_run = not args.execute
    today   = date.today()
    dates   = [(today + timedelta(d)).strftime("%-m/%-d/%Y") for d in range(1, args.days + 1)]

    print(f"\nFetching schedule for {dates[0]} → {dates[-1]}...")
    analyses = []
    with browser_session(headless=True) as page:
        for d in dates:
            items = fetch_schedule(d, d, page=page)
            analyses.append(analyse_day(d, items))

    changes = plan_changes(analyses)
    print_summary(analyses, changes)

    if dry_run:
        print("  Dry run — no changes made.")
        print("  Run with --execute to apply changes.\n")
    else:
        print(f"  Executing {len(changes)} change(s)...")
        execute_changes(changes, dry_run=False)
        print("\n  Done.\n")


if __name__ == "__main__":
    main()

File: ts/src/jobs/fixImbalance.ts
/**
 * AI/Intermediate rebalancer — port of `fix_imbalance.py`.
 *
 * Over the next N days: keep at most 1 Advanced Intermediate per day (cancel
 * extras, but NEVER one with registered members), pair each AI with a
 * simultaneous Intermediate on another court, and top Intermediate up to a daily
 * target. Book/cancel go through `courtreserve-api` (no browser).
 */
import 'dotenv/config'
import { resolve } from 'node:path'
import { CourtReserveClient } from '../cr/client'
import type { ScheduleItem } from '../cr/types'
import { loadPolicy, type Policy } from '../policy'
import { NaiveDateTime } from '../datetime'

const INT_TARGET = 2
const INT_TIMES = [9, 11, 13, 15]
const MAX_CONCURRENT_COURTS = 3

interface Courts {
  [id: string]: { number: number; label?: string }
}
interface ApprovedEvents {
  [id: string]: { name: string; level: string }
}

interface Ctx {
  courts: Courts
  courtIdsByPref: string[] // court 4 first, then 1,2,3
  aiEventId: number
  intEventId: number
  approved: ApprovedEvents
  policy: Policy
}

function buildCtx(policy: Policy): Ctx {
  const courts = policy.courts as unknown as Courts
  const approved = policy.approved_events as unknown as ApprovedEvents
  const byLevel: Record<string, string> = {}
  for (const [eid, info] of Object.entries(approved)) byLevel[info.level] = eid
  const courtIdsByPref = Object.keys(courts).sort((a, b) => {
    const an = courts[a].number === 4 ? 0 : 1
    const bn = courts[b].number === 4 ? 0 : 1
    return an - bn || courts[a].number - courts[b].number
  })
  return {
    courts,
    courtIdsByPref,
    aiEventId: Number(byLevel['Advanced Intermediate']),
    intEventId: Number(byLevel['Intermediate']),
    approved,
    policy,
  }
}

function levelOf(ctx: Ctx, item: ScheduleItem): string | null {
  const eid = String(item.EventId ?? '')
  return eid in ctx.approved ? ctx.approved[eid].level : null
}

/** First free court id at [start,end], respecting the 3-court concurrent max. */
function findFreeCourt(
  ctx: Ctx,
  items: ScheduleItem[],
  start: NaiveDateTime,
  end: NaiveDateTime,
): number | null {
  const occupied = new Set<number>()
  for (const item of items) {
    if (!item.StartDateTime || !item.EndDateTime) continue
    const s = NaiveDateTime.fromISO(item.StartDateTime)
    const e = NaiveDateTime.fromISO(item.EndDateTime)
    if (s.ms < end.ms && e.ms > start.ms) {
      const courtsStr = String(item.Courts ?? '')
      for (const cid of Object.keys(ctx.courts)) {
        if (courtsStr.includes(`#${ctx.courts[cid].number}`) || courtsStr.includes(cid)) {
          occupied.add(Number(cid))
        }
      }
    }
  }
  if (occupied.size >= MAX_CONCURRENT_COURTS) return null
  for (const cid of ctx.courtIdsByPref) {
    if (!occupied.has(Number(cid))) return Number(cid)
  }
  return null
}

interface DayEvent {
  event_name: string
  start: NaiveDateTime
  end: NaiveDateTime
  courts: string
  members: number
  occurrence_id: number | undefined
  event_id: number | undefined
}

export interface DayAnalysis {
  date_str: string
  dow: string
  ai_events: DayEvent[]
  int_events: DayEvent[]
  all_items: ScheduleItem[]
}

export function analyseDay(ctx: Ctx, dateStr: string, items: ScheduleItem[]): DayAnalysis {
  const dt = NaiveDateTime.parseDate(dateStr)
  const ymd = dt.formatYmd()
  const ai: DayEvent[] = []
  const int: DayEvent[] = []
  for (const item of items) {
    const level = levelOf(ctx, item)
    if (!item.StartDateTime) continue
    const start = NaiveDateTime.fromISO(item.StartDateTime)
    if (start.formatYmd() !== ymd) continue
    const name = String(item.EventName ?? '').toLowerCase()
    const resType = String((item as { ReservationType?: string }).ReservationType ?? '').toLowerCase()
    if (name.includes('cancel') || resType.includes('cancel')) continue
    const rec: DayEvent = {
      event_name: String(item.EventName ?? ''),
      start,
      end: NaiveDateTime.fromISO(item.EndDateTime as string),
      courts: String(item.Courts ?? ''),
      members: Number((item as { MembersCount?: number }).MembersCount ?? 0),
      occurrence_id: item.Id,
      event_id: item.EventId,
    }
    if (level === 'Advanced Intermediate') ai.push(rec)
    else if (level === 'Intermediate') int.push(rec)
  }
  const byStart = (a: DayEvent, b: DayEvent) => a.start.ms - b.start.ms
  return {
    date_str: dateStr,
    dow: dt.weekdayName(),
    ai_events: ai.sort(byStart),
    int_events: int.sort(byStart),
    all_items: items,
  }
}

export type Change =
  | {
      action: 'cancel'
      date_str: string
      dow: string
      event_id: number
      occurrence_id: number
      description: string
    }
  | {
      action: 'book'
      date_str: string
      dow: string
      event_id: number
      start_time: string
      end_time: string
      court_id: number
      court_num: number
      description: string
    }

const ampm = (dt: NaiveDateTime) => dt.formatTime().replace(':00 ', '').replace(' ', '') // "9AM"

export function planChanges(ctx: Ctx, analyses: DayAnalysis[], log: (m: string) => void = () => {}): Change[] {
  const changes: Change[] = []
  const win = ctx.policy.operating_windows

  for (const day of analyses) {
    const { date_str: dateStr, dow } = day
    const aiEvents = day.ai_events
    // Working copies mutated as we book, so later passes respect placements.
    const intEvents: Array<{ start: NaiveDateTime; members: number }> = day.int_events.map((e) => ({
      start: e.start,
      members: e.members,
    }))
    const itemsForDay = day.all_items
    const dayDt = NaiveDateTime.parseDate(dateStr)
    const ymd = dayDt.formatYmd()
    const window = dow !== 'Saturday' && dow !== 'Sunday' ? win.weekday : win.weekend
    const winStartH = Number(window.start.split(':')[0])
    const winEndH = Number(window.end.split(':')[0])

    // ── AI: keep at most 1 ────────────────────────────────────────────────────
    if (aiEvents.length > 1) {
      const withMembers = aiEvents.filter((e) => e.members > 0)
      const keep = withMembers[0] ?? aiEvents[0]
      for (const ev of aiEvents) {
        if (ev === keep) continue
        if (ev.members > 0) {
          log(`  ${dateStr}  AI at ${ampm(ev.start)} has ${ev.members} members — SKIPPING cancel`)
          continue
        }
        if (ev.event_id == null || ev.occurrence_id == null) continue
        changes.push({
          action: 'cancel',
          date_str: dateStr,
          dow,
          event_id: ev.event_id,
          occurrence_id: ev.occurrence_id,
          description: `Cancel extra AI at ${ampm(ev.start)} (${ev.courts})`,
        })
      }
    }

    // ── Pair Intermediate with each AI session ───────────────────────────────
    for (const aiEv of aiEvents) {
      const aiHour = aiEv.start.hour
      if (intEvents.some((ie) => ie.start.hour === aiHour)) continue
      if (aiHour < winStartH || aiHour + 2 > winEndH) {
        log(`  ${dateStr}  AI at ${ampm(aiEv.start)}: Intermediate pairing skipped — outside window`)
        continue
      }
      const startDt = NaiveDateTime.fromYMDHM(ymd, `${String(aiHour).padStart(2, '0')}:00`)
      const endDt = startDt.addHours(2)
      const courtId = findFreeCourt(ctx, itemsForDay, startDt, endDt)
      if (courtId === null) {
        log(`  ${dateStr}  AI at ${ampm(aiEv.start)}: Intermediate pairing skipped — no free court`)
        continue
      }
      const courtNum = ctx.courts[String(courtId)].number
      changes.push({
        action: 'book',
        date_str: dateStr,
        dow,
        event_id: ctx.intEventId,
        start_time: startDt.formatTime(),
        end_time: endDt.formatTime(),
        court_id: courtId,
        court_num: courtNum,
        description: `Book Intermediate ${ampm(startDt)}–${ampm(endDt)} Court #${courtNum} (paired with AI)`,
      })
      itemsForDay.push(synthItem(ctx, ctx.intEventId, startDt, endDt, courtId))
      intEvents.push({ start: startDt, members: 0 })
    }

    // ── Intermediate: add up to target ───────────────────────────────────────
    let needed = INT_TARGET - intEvents.length
    for (const hour of INT_TIMES) {
      if (needed <= 0) break
      if (hour < winStartH || hour + 2 > winEndH) continue
      const startDt = NaiveDateTime.fromYMDHM(ymd, `${String(hour).padStart(2, '0')}:00`)
      const endDt = startDt.addHours(2)
      const already = intEvents.some(
        (e) => Math.abs((e.start.hour * 60 + e.start.minute) - hour * 60) < 120,
      )
      if (already) continue
      const courtId = findFreeCourt(ctx, itemsForDay, startDt, endDt)
      if (courtId === null) continue
      const courtNum = ctx.courts[String(courtId)].number
      changes.push({
        action: 'book',
        date_str: dateStr,
        dow,
        event_id: ctx.intEventId,
        start_time: startDt.formatTime(),
        end_time: endDt.formatTime(),
        court_id: courtId,
        court_num: courtNum,
        description: `Book Intermediate ${ampm(startDt)}–${ampm(endDt)} Court #${courtNum}`,
      })
      itemsForDay.push(synthItem(ctx, ctx.intEventId, startDt, endDt, courtId))
      intEvents.push({ start: startDt, members: 0 })
      needed -= 1
    }
  }
  return changes
}

function synthItem(
  ctx: Ctx,
  eventId: number,
  start: NaiveDateTime,
  end: NaiveDateTime,
  courtId: number,
): ScheduleItem {
  return {
    EventId: eventId,
    StartDateTime: `${start.formatYmd()}T${start.formatHm()}:00`,
    EndDateTime: `${end.formatYmd()}T${end.formatHm()}:00`,
    Courts: `#${ctx.courts[String(courtId)].number}`,
    CourtId: String(courtId),
    MembersCount: 0,
  } as ScheduleItem
}

export async function executeChanges(
  cr: CourtReserveClient,
  changes: Change[],
  log: (m: string) => void = () => {},
): Promise<void> {
  for (const c of changes.filter((c): c is Extract<Change, { action: 'cancel' }> => c.action === 'cancel')) {
    log(`Cancelling ${c.description} (occ=${c.occurrence_id})`)
    try {
      await cr.cancel({ res_id: String(c.occurrence_id) })
      log(`  ✅ ${c.description}`)
    } catch (e) {
      log(`  ❌ FAILED: ${e instanceof Error ? e.message : String(e)}  ${c.description}`)
    }
  }
  for (const c of changes.filter((c): c is Extract<Change, { action: 'book' }> => c.action === 'book')) {
    log(`Booking ${c.description}`)
    try {
      await cr.book({
        event_id: String(c.event_id),
        date: c.date_str,
        start_time: c.start_time,
        end_time: c.end_time,
        court_id: String(c.court_id),
        dry_run: false,
      })
      log(`  ✅ ${c.description}`)
    } catch (e) {
      log(`  ❌ FAILED: ${e instanceof Error ? e.message : String(e)}  ${c.description}`)
    }
  }
}

export async function runFixImbalance(
  cr: CourtReserveClient,
  opts: { days?: number; execute?: boolean; today?: Date; policy?: Policy; log?: (m: string) => void } = {},
): Promise<Change[]> {
  const log = opts.log ?? (() => {})
  const policy = opts.policy ?? loadPolicy()
  const ctx = buildCtx(policy)
  const days = opts.days ?? 14
  const today = opts.today ?? new Date()

  const analyses: DayAnalysis[] = []
  for (let d = 1; d <= days; d++) {
    const day = new Date(today)
    day.setDate(day.getDate() + d)
    const dateStr = `${day.getMonth() + 1}/${day.getDate()}/${day.getFullYear()}`
    const items = await cr.schedule(dateStr, dateStr)
    analyses.push(analyseDay(ctx, dateStr, items))
  }

  const changes = planChanges(ctx, analyses, log)
  log(`Planned ${changes.length} change(s).`)
  for (const c of changes) {
    log(`  ${c.action === 'cancel' ? '❌ CANCEL' : '✅ BOOK'}  ${c.dow.slice(0, 3)} ${c.date_str}  ${c.description}`)
  }

  if (opts.execute) {
    log(`Executing ${changes.length} change(s)…`)
    await executeChanges(cr, changes, log)
  } else {
    log('Dry run — no changes made. Pass --execute to apply.')
  }
  return changes
}

// exported for tests
export { buildCtx, findFreeCourt }

const isMain = process.argv[1] && import.meta.url === `file://${process.argv[1]}`
if (isMain) {
  const argv = process.argv.slice(2)
  const daysIdx = argv.indexOf('--days')
  const cr = new CourtReserveClient(
    process.env.CRAPI_URL ?? 'http://localhost:8787',
    process.env.CRAPI_KEY ?? '',
  )
  await runFixImbalance(cr, {
    days: daysIdx >= 0 ? Number(argv[daysIdx + 1]) : 14,
    execute: argv.includes('--execute'),
    log: (m) => console.log(m),
  })
}
